### PR TITLE
Added conversion between lists and arrays

### PR DESCRIPTION
**Pull request by jimbxb**
Also added a `word_size` constant, slight changes to docs and a minor change to `range`

### CODE REVIEW

File: wybelibs/wybe/list.wybe [hunk 1]
```diff
@@ -129,6 +121,35 @@ pub def println(printer:{resource}(T), xs:_(T)) use !io {
     !nl
 }
 
+
+## Conversion procedures
+
+# Conversion an array into a list
```
Comment (pschachte) [inline wybelibs/wybe/list.wybe:127]: Typo

File: wybelibs/wybe/list.wybe [hunk 2]
```diff
@@ -129,6 +121,35 @@ pub def println(printer:{resource}(T), xs:_(T)) use !io {
...
+pub def list(ar:array(T)):_(T) = ls where {
+    ?ls = []
+    for ?x in ar {
+        ?ls = [x | ls]
+    }
+    reverse(ls, ?ls)
+}
+
```
Comment (pschachte) [inline wybelibs/wybe/list.wybe:135]: It should be possible to do this more efficiently using recursion and making use of Chris's TCMC optimisation.

File: wybelibs/wybe/machine_word.wybe [hunk 3]
```diff
@@ -4,3 +4,7 @@
 pragma no_standard_library # Standard library can't depend on itself!
 
 representation is address
+
+use wybe.int
+
+pub def word_size:int = 8
```
Comment (pschachte) [inline wybelibs/wybe/machine_word.wybe:10]: It would be better to call it word_size_bytes, and also have a word_size_bits constant.

**Review by pschachte — APPROVED** [review]

File: wybelibs/wybe/list.wybe (hunk 2 above)
Comment (jimbxb) [reply wybelibs/wybe/list.wybe:135]: True

Comment (jimbxb) [reply wybelibs/wybe/list.wybe:135]: I have tried making a TCO version of the proc as follows:

```
pub def list(ar:array(T)):list(T) = ls where {
    if { [?x | ?ar] = ar ::
        ?ls = list(ar)
        [x | ls] = ?ls 
       | else :: 
        ?ls = []
    }
}
```

However I think that multi-specz is getting in the way

```
list > public (1 calls)
0: opt.list<0>
list(ar##0:wybe.array(T), ?#result##0:wybe.list(T))<{}; {}>:
  AliasPairs: []
  InterestingCallProperties: [InterestingUnaliased 0]
  MultiSpeczDepInfo: [(0,(wybe.array.[|]<0>,fromList [NonAliasedParamCond 2 [0]])),(1,(opt.list<0>,fromList [NonAliasedParamCond 0 []]))]
    wybe.array.[|]<0>(?x##0:T, ?ar##1:wybe.array(T), ~ar##0:wybe.array(T), ?tmp#3##0:wybe.bool) #0 @opt:16:10
    case ~tmp#3##0:wybe.bool of
    0:
        foreign llvm move(0:wybe.list(T), ?#result##0:wybe.list(T)) @opt:15:5

    1:
        opt.list<0>[410bae77d3](~ar##1:wybe.array(T), ?tmp#0##0:wybe.list(T)) #1 @opt:17:15
        foreign lpvm alloc(16:wybe.int, ?tmp#6##0:wybe.list(T)) @list:9:27
        foreign lpvm mutate(~tmp#6##0:wybe.list(T), ?tmp#7##0:wybe.list(T), 0:wybe.int, 1:wybe.int, 16:wybe.int, 0:wybe.int, ~x##0:T) @list:9:27
        foreign lpvm mutate(~tmp#7##0:wybe.list(T), ?#result##0:wybe.list(T), 8:wybe.int, 1:wybe.int, 16:wybe.int, 0:wybe.int, ~tmp#0##0:wybe.list(T)) @list:9:27
```
And with `-xno-multi-specz`:
```list > public (1 calls)
0: opt.list<0>
list(ar##0:wybe.array(T), outByReference #result##0:wybe.list(T))<{}; {}>:
  AliasPairs: []
  InterestingCallProperties: [InterestingUnaliased 0]
  MultiSpeczDepInfo: [(0,(wybe.array.[|]<0>,fromList [NonAliasedParamCond 2 [0]])),(1,(opt.list<0>,fromList [NonAliasedParamCond 0 []]))]
    wybe.array.[|]<0>(?x##0:T, ?ar##1:wybe.array(T), ~ar##0:wybe.array(T), ?tmp#3##0:wybe.bool) #0 @opt:16:10
    case ~tmp#3##0:wybe.bool of
    0:
        foreign llvm move(0:wybe.list(T), ?#result##0:wybe.list(T)) @opt:15:5

    1:
        foreign lpvm alloc(16:wybe.int, ?tmp#6##0:wybe.list(T)) @list:9:27
        foreign lpvm mutate(~tmp#6##0:wybe.list(T), ?tmp#7##0:wybe.list(T), 0:wybe.int, 1:wybe.int, 16:wybe.int, 0:wybe.int, ~x##0:T) @list:9:27
        opt.list<0>(~ar##1:wybe.array(T), outByReference tmp#0##0:wybe.list(T)) #1 @opt:17:15
        foreign lpvm mutate(~tmp#7##0:wybe.list(T), ?#result##0:wybe.list(T), 8:wybe.int, 1:wybe.int, 16:wybe.int, 0:wybe.int, ~takeReference tmp#0##0:wybe.list(T)) @list:9:27
```

Comment (jimbxb) [reply wybelibs/wybe/list.wybe:135]: I think TCMC needs to learn about multi-specz

**Review by pschachte — APPROVED** [review]

**Comment by pschachte** [comment]
Can you please put in an issue regarding multiple specialisation?  The TCMC pass should be run on code generated by multiple specialisation, too.

**Pull request merged**